### DISCUSSION TITLE
怎样异步运行Server

### DISCUSSION THREAD

**Discussion by OstrichRunning** [Tencent/tquic#291] [General]
我看demo上运行一个Server就进入一个loop，如果我想同时启动v4和v6两个socket，我该怎样做？
当我使用tokio::spawn(async move {} 时
报下面的错误
`(dyn Http3Handler + 'static)` cannot be shared between threads safely

麻烦帮看看下，如何能异步启动服务

**Comment by glime-ai** [comment]
要同时启动IPv4和IPv6两个socket，并且异步启动服务，你可以参考以下步骤：

1. **创建两个不同的SocketAddr**，一个用于IPv4，一个用于IPv6。
2. **使用tokio::spawn**来启动两个不同的任务，每个任务监听一个socket。
3. 确保你的`Http3Handler`实现了`Send`和`Sync` trait，以便它可以在多个线程之间安全地共享。

以下是一个简单的示例代码：

```rust
use tokio::net::TcpListener;
use tokio::runtime::Runtime;
use std::sync::Arc;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let ipv4_addr = "127.0.0.1:8080".parse()?;
    let ipv6_addr = "[::1]:8080".parse()?;

    let listener_v4 = TcpListener::bind(ipv4_addr).await?;
    let listener_v6 = TcpListener::bind(ipv6_addr).await?;

    let handler = Arc::new(MyHttp3Handler::new());

    let handler_v4 = handler.clone();
    let handler_v6 = handler.clone();

    tokio::spawn(async move {
        loop {
            let (socket, _) = listener_v4.accept().await.unwrap();
            let handler = handler_v4.clone();
            tokio::spawn(async move {
                // 处理连接
                handler.handle(socket).await;
            });
        }
    });

    tokio::spawn(async move {
        loop {
            let (socket, _) = listener_v6.accept().await.unwrap();
            let handler = handler_v6.clone();
            tokio::spawn(async move {
                // 处理连接
                handler.handle(socket).await;
            });
        }
    });

    // 保持主线程运行
    loop {
        tokio::time::sleep(tokio::time::Duration::from_secs(60)).await;
    }
}

struct MyHttp3Handler;

impl MyHttp3Handler {
    fn new() -> Self {
        MyHttp3Handler
    }

    async fn handle(&self, socket: tokio::net::TcpStream) {
        // 处理连接的逻辑
    }
}

// 确保你的Handler实现了Send和Sync
unsafe impl Send for MyHttp3Handler {}
unsafe impl Sync for MyHttp3Handler {}
```

在这个示例中，我们创建了两个`TcpListener`，分别监听IPv4和IPv6地址。然后，我们使用`tokio::spawn`启动两个异步任务，每个任务处理一个监听器的连接。为了确保`Http3Handler`可以在多个线程之间共享，我们使用了`Arc`并确保它实现了`Send`和`Sync`。

This comment was generated by [Glime](https://glimelab.ai/?utm_source=44520fce49ba43798485a628d6928e22&utm_medium=github).

**Comment by iyangsj** [comment]
目前暂未提供异步Rust风格的接口，后续版本会支持

**Comment by iyangsj** [comment]
> 我看demo上运行一个Server就进入一个loop，如果我想同时启动v4和v6两个socket，我该怎样做？

示例Server可支持处理IPv4/IPv6双栈流量，例如指定 `--listen [::]:443` 参数，可监听来自IPv6/IPv4双栈的443端口的UDP报文。

你也可以你的Server程序中，分别为IPv4/IPv6创建多个套接字，并分别通过这些套接字接收和发送UDP报文。

**Comment by OstrichRunning** [comment]
<img width="200" alt="image" src="https://github.com/user-attachments/assets/c1baa9e7-2432-4035-acec-b51e75727efd">

我这边生成静态库，swift调用函数quic_endpoint_connect返回-101错误，咨询一下什么地方可以查询错误原因？像这类错误，我该如何调试（静态库）？

**Reply by iyangsj** [reply]
quic接口错误码见 https://github.com/Tencent/tquic/blob/develop/src/error.rs#L185
先确认下接口使用是否正确，也可以把日志输出及代码贴出看看

**Reply by OstrichRunning** [reply]
quic_endpoint_recv这个调用和quic_endpoint_connect这个调用是有先后顺序吗？

**Reply by OstrichRunning** [reply]
我调用 quic_endpoint_new 后 紧接着就调用 quic_endpoint_connect 就报错了
<img width="898" alt="image" src="https://github.com/user-attachments/assets/642042f0-4684-4a09-9765-a7ea9ebce383">
这个是swift侧的调用

**Reply by OstrichRunning** [reply]
还有一个问题是：我看rust代码接收数据利用了poll，我在swift只能调用在while循环调用quic_endpoint_recv来接收数据吗？还是说是有类似poll的功能，我看swift有个库SwiftNIO，是否可以利用这个来避免直接的循环呢
